### PR TITLE
[Merged by Bors] - feat: define `IsLittleOTVS`

### PR DESCRIPTION
**Pull request by urkud**
The key theorem here is `isLittleOTVS_iff_isLittleO`, which enables us in a future PR to redefine `HasFDerivAtFilter` to not be dependent on a choice of norm, making it possible to _state_ results about derivatives of types with no canonical norm.
This gives us a middle ground of being able to state `Matrix.hasDerivAt_add`, which is better than not being able to state it, but not as good as being able to reuse the global `hasDerivAt_add`.

Later on, we can try to generalize results like `IsLittleO.add` to `IsLittleOTVS`, which would help with this reuse.

[Zulip thread](https://leanprover.zulipchat.com/#narrow/stream/116395-maths/topic/generalizing.20deriv.20to.20TVS/near/359284921)

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by eric-wieser** [comment]
I think this is probably reviewable now; ideally we'd generalize the equivalence between this definition and the existing one to seminormed spaces somehow, but that's not relevant to the subsequent `fderiv` generalization anyway.

File: Mathlib/Analysis/Asymptotics/TVS.lean [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+import Mathlib.Topology.Instances.ENNReal
+
+/-!
+# Asymptotics in a Topological Vector Space
+
+This file defines `Asymptotics.IsLittleOTVS` as a generalization of `Asymptotics.IsLittleO` from
+normed spaces to toplogical spaces.
+
+This frees the user from having to chose a canonical norm, at the expense of having to pick a
+specific base ring.
```
Comment (eric-wieser) [inline Mathlib/Analysis/Asymptotics/TVS.lean:20]: @urkud, can you write something here comparing to https://en.wikipedia.org/wiki/Fr%C3%A9chet_derivative#Generalization_to_topological_vector_spaces ?

File: Mathlib/Analysis/Convex/EGauge.lean [hunk 2]
```diff
@@ -177,27 +182,28 @@ lemma div_le_egauge_ball (r : ℝ≥0) (x : E) : ‖x‖₊ / r ≤ egauge 𝕜
...
-
-section SeminormedAddCommGroup
-
-variable {𝕜 : Type*} [NormedField 𝕜] {E : Type*}
-    [NormedAddCommGroup E] [NormedSpace 𝕜 E] {c : 𝕜} {x : E} {r : ℝ≥0}
+variable {𝕜}
+variable {c : 𝕜} {x : E} {r : ℝ≥0}
 
-lemma egauge_ball_le_of_one_lt_norm (hc : 1 < ‖c‖) (h₀ : r ≠ 0 ∨ x ≠ 0) :
+lemma egauge_ball_le_of_one_lt_norm (hc : 1 < ‖c‖) (h₀ : r ≠ 0 ∨ ‖x‖ ≠ 0) :
```
Comment (eric-wieser) [inline Mathlib/Analysis/Convex/EGauge.lean:188]: Apparently I falsely concluded that something was wrong with `egauge`; the fact that you made this work withotu changing its definition is great!

File: Mathlib/Analysis/Asymptotics/TVS.lean [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+
+This definition was added to the library in order to migrate Fréchet derivatives
+from normed vector spaces to topological vector spaces.
+The definition is motivated by
+https://en.wikipedia.org/wiki/Fr%C3%A9chet_derivative#Generalization_to_topological_vector_spaces
+but the definition there doesn't work for topological vector spaces over general normed fields.
+For the discussion that lead to the current choice of the definition, see
+https://leanprover.zulipchat.com/#narrow/channel/116395-maths/topic/generalizing.20deriv.20to.20TVS
```
Comment (eric-wieser) [inline Mathlib/Analysis/Asymptotics/TVS.lean:28]:
```suggestion
[This Zulip discussion](https://leanprover.zulipchat.com/#narrow/channel/116395-maths/topic/generalizing.20deriv.20to.20TVS)
led to the current choice of the definition.
```
(led not lead, and named links to make this look better in the docs)

File: Mathlib/Analysis/Asymptotics/TVS.lean (hunk 1 above)
Comment (urkud) [reply Mathlib/Analysis/Asymptotics/TVS.lean:20]: I wrote something. Feel free to update.

**Review by eric-wieser — APPROVED** [review]
I edited the PR description to give an explicit example of how this is an improvement over the status quo, while also only being the first step.

File: Mathlib/Analysis/Asymptotics/TVS.lean [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+the choice of norm being part of the statement.
+-/
+
+open Set Filter Asymptotics Metric
+open scoped Topology Pointwise ENNReal NNReal
+
+namespace Asymptotics
+
+/-- `IsLittleOTVS 𝕜 f g l` is a generalization of `f =o[l] g` (`IsLittleO f g l`) that works in
+topological `𝕜`-vector spaces. -/
```
Comment (sgouezel) [inline Mathlib/Analysis/Asymptotics/TVS.lean:50]: Could you expand the docstring, by explaining what the definition says?

**Review by sgouezel — COMMENTED** [review]
This looks very nice, thanks!

Could you explain the choice of definition in the file-level docstring, instead of just referring to a Zulip discussion? In particular, it would be nice to explain why you define `IsLittleOTVS` but not `IsBigOTVS` (is the latter meaningless?).

File: Mathlib/Topology/Instances/ENNReal.lean [hunk 5]
```diff
@@ -76,7 +76,7 @@ theorem tendsto_coe {f : Filter α} {m : α → ℝ≥0} {a : ℝ≥0} :
     Tendsto (fun a => (m a : ℝ≥0∞)) f (𝓝 ↑a) ↔ Tendsto m f (𝓝 a) :=
   isEmbedding_coe.tendsto_nhds_iff.symm
 
-@[fun_prop]
+@[continuity, fun_prop]
```
Comment (j-loreaux) [inline Mathlib/Topology/Instances/ENNReal.lean:79]: What is still using `continuity` that you needed to add this?

**Review by j-loreaux — COMMENTED** [review]
maintainer delegate

File: Mathlib/Topology/Instances/ENNReal.lean (hunk 5 above)
Comment (urkud) [reply Mathlib/Topology/Instances/ENNReal.lean:79]: I guess, I started this PR a long time ago when `fun_prop` was rarely used in Mathlib. Let me try to revert this line.

Comment (urkud) [reply Mathlib/Topology/Instances/ENNReal.lean:79]:
```suggestion
@[fun_prop]
```

**Comment by sgouezel** [comment]
bors r+
Thanks!

**Pull request closed without merging**